Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellCollection.cs

```diff
@@ -0,0 +1,164 @@
+        public virtual void AddRange(params DataGridViewCell[] dataGridViewCells)
+        {
+            Contract.Requires(dataGridViewCells != null);
+        }
+
+        // <summary>
+        // Clears all cells from the collection.
+        // </summary>
+        // <exception cref="T:System.InvalidOperationException">The row that owns this <see cref="T:System.Windows.Forms.DataGridViewCellCollection"/> already belongs to a <see cref="T:System.Windows.Forms.DataGridView"/> control.</exception>
+        // public virtual void Clear()
+        
+        // <summary>
+        // Copies the entire collection of cells into an array at a specified location within the array.
+        // </summary>
+        // <param name="array">The destination array to which the contents will be copied.</param><param name="index">The index of the element in <paramref name="array"/> at which to start copying.</param>
+        // public void CopyTo(DataGridViewCell[] array, int index)
+        
+        // <summary>
+        // Determines whether the specified cell is contained in the collection.
+        // </summary>
+        // 
+        // <returns>
+        // true if <paramref name="dataGridViewCell"/> is in the collection; otherwise, false.
+        // </returns>
+        // <param name="dataGridViewCell">A <see cref="T:System.Windows.Forms.DataGridViewCell"/> to locate in the collection.</param>
+        // public virtual bool Contains(DataGridViewCell dataGridViewCell)
+        
+        // <summary>
+        // Returns the index of the specified cell.
+        // </summary>
+        // 
+        // <returns>
+        // The zero-based index of the value of <paramref name="dataGridViewCell"/> parameter, if it is found in the collection; otherwise, -1.
+        // </returns>
+        // <param name="dataGridViewCell">The cell to locate in the collection.</param>
+        // public int IndexOf(DataGridViewCell dataGridViewCell)
+        
+        /// <summary>
+        /// Inserts a cell into the collection at the specified index.
+        /// </summary>
+        /// <param name="index">The zero-based index at which to place <paramref name="dataGridViewCell"/>.</param><param name="dataGridViewCell">The <see cref="T:System.Windows.Forms.DataGridViewCell"/> to insert.</param><exception cref="T:System.InvalidOperationException">The row that owns this <see cref="T:System.Windows.Forms.DataGridViewCellCollection"/> already belongs to a <see cref="T:System.Windows.Forms.DataGridView"/> control.-or-<paramref name="dataGridViewCell"/> already belongs to a <see cref="T:System.Windows.Forms.DataGridViewRow"/>.</exception>
+        public virtual void Insert(int index, DataGridViewCell dataGridViewCell)
+        {
+            Contract.Requires(index >= 0);
+        }
```

Comment:
Two additional requirements could be added:

`Contract.Requires(dataGridViewCell != null);`
`Contract.Requires(dataGridViewCell.OwningRow == null);`